language/stdlib/modules/gas_schedule.mvir

```diff
@@ -0,0 +1,324 @@
+// The gas schedule keeps two separate schedules for the gas:
+// * The instruction_schedule: This holds the gas for each bytecode instruction.
+// * The native_schedule: This holds the gas for used (per-byte operated over) for each native
+//   function.
+// A couple notes:
+// 1. The functions that update the cost table are meant to be public -- the
+//    association will send a normal transaction to these functions in order
+//    to update any costs, or to add any new costs.
+// 2. In the case that e.g. an instruction is deleted from the bytecode,
+//    that part of the cost schedule needs to be filled in by either a
+//    dummy cost, or by the cost of another instruction (and that
+//    instruction must map to that slot).
+// 3. The initialization of the module will publish the instruction table
+//    to the association address, and will preload the vector with the
+//    gas schedule for instructions. The VM will then load this into memory at startup.
+module GasSchedule {
+    import 0x0.Vector;
+
+    // The gas cost for each instruction is represented using two amounts;
+    // one for the cpu, and the other for storage.
+    struct Cost {
+      cpu: u64,
+      storage: u64,
+    }
+
+    resource T {
+        instruction_schedule: Vector.T<Self.Cost>,
+        native_schedule: Vector.T<Self.Cost>,
+    }
+
+    // Initialize the table under the association account
+    initialize() {
+        let instruction_table: Vector.T<Self.Cost>;
+        assert(get_txn_sender() == 0xA550C18, 0);
+        instruction_table = Vector.empty<Self.Cost>();
+
+        // NB: DO NOT change the order of these pushes to the vector.
+        Vector.push_back<Self.Cost>(
```

Comment:
This is the order in which the bytecodes are serialized -- e.g. Add has OPCODE 0x2 (or something), therefore the cost for the Add instruction would be at offset 1 (since the vector is 0-indexed and the instructions are 1-indexed).